### PR TITLE
Monte carlo packets

### PR DESCRIPTION
**Pull request by ffoucart**
## Proposed changes

Add packets for MC evolutions, and ability to evolve packets along a geodesic

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

Very preliminary; packets cannot at this point be included in actual evolution...

### CODE REVIEW

File: tests/Unit/Evolution/Particles/MonteCarlo/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+set(LIBRARY_SOURCES
+  Test_MonteCarloPacket.cpp
+  )
+
+add_test_library(
+  ${LIBRARY}
+  "Evolution/Particles/MonteCarlo"
+  "${LIBRARY_SOURCES}"
+  "Boost::boost;Evolution;Utilities"
```
Comment (kidder) [inline tests/Unit/Evolution/Particles/MonteCarlo/CMakeLists.txt]: you are missing `MonteCarlo` in this list

Better is to update this to the following (which is what we are slowly doing to all test directories):
- Change line 14 to just be an empty list `""`
- Add the following to the end of the file:
```
target_link_libraries(
  ${LIBRARY}
  PRIVATE
  Monte Carlo
  )
```

**Comment by ffoucart** [comment]
@kidder : The code for the evolution of a single packet for a time step along a geodesic seems to pass the standard tests now. I'm not sure what a more appropriate test would be here; the one I implemented is obviously simplistic. I could write a python function reproducing the whole second-order time stepping, and compare results with random inputs, I guess? It's not clear to me whether we will actually perform time steps using this method, or incorporate the time-stepping within another part of the code once we try to actually move packets on a grid.

Apart from that, I think this PR is ready for review (and, obviously, fairly low priority...).

File: src/Evolution/Particles/MonteCarlo/EvolveMonteCarloPackets.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    const gsl::not_null<Packet*> packet, const double time_step,
+    const Scalar<DataVector>& lapse,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& shift,
+    const tnsr::i<DataVector, 3, Frame::Inertial>& d_lapse,
+    const tnsr::iJ<DataVector, 3, Frame::Inertial>& d_shift,
+    const tnsr::iJJ<DataVector, 3, Frame::Inertial>& d_inv_spatial_metric,
+    const tnsr::II<DataVector, 3, Frame::Inertial>& inv_spatial_metric,
+    const size_t& closest_point_index) {
+  // Temporary variables
+  std::vector<double> dpdt(3, 0);
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/EvolveMonteCarloPackets.cpp]: Can you please change these to `std::array<double, 3>`?

File: src/Evolution/Particles/MonteCarlo/EvolveMonteCarloPackets.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    const tnsr::I<DataVector, 3, Frame::Inertial>& mesh_coordinates,
+    const Scalar<DataVector>& lapse,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& shift,
+    const tnsr::i<DataVector, 3, Frame::Inertial>& d_lapse,
+    const tnsr::iJ<DataVector, 3, Frame::Inertial>& d_shift,
+    const tnsr::iJJ<DataVector, 3, Frame::Inertial>& d_inv_spatial_metric,
+    const tnsr::II<DataVector, 3, Frame::Inertial>& inv_spatial_metric) {
+  // Mesh information. Currently assumes uniform grid without map
+  const Index<3>& extents = mesh.extents();
+  const std::vector<double> bottom_coord_mesh{mesh_coordinates.get(0)[0],
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/EvolveMonteCarloPackets.cpp]: Can you please change this and the next `std::vector` to `std::array<double, 3>`?

File: src/Evolution/Particles/MonteCarlo/EvolveMonteCarloPackets.cpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  const std::vector<double> dx_mesh{
+      mesh_coordinates.get(0)[1] - bottom_coord_mesh[0],
+      mesh_coordinates.get(1)[extents[0]] - bottom_coord_mesh[1],
+      mesh_coordinates.get(2)[extents[0] * extents[1]] - bottom_coord_mesh[2]};
+
+  // Loop over packets
+  for (size_t p = 0; p < packets->size(); p++) {
+    Packet& packet = (*packets)[p];
+    // Find closest grid point to packet
+    std::vector<size_t> closest_point_index_3d(3, 0);
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/EvolveMonteCarloPackets.cpp]: Can you please change these to `std::array<size_t, 3>`?

File: src/Evolution/Particles/MonteCarlo/EvolveMonteCarloPackets.hpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+namespace MonteCarlo {
+
+struct Packet;
+
+namespace detail {
+
+// Time derivative of the spatial component of the momentum one-form on a null
+// geodesic
+void time_derivative_momentum_geodesic(
+    const gsl::not_null<std::vector<double>*> dt_momentum, const Packet& packet,
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/EvolveMonteCarloPackets.hpp]: When you rebase, clang-tidy may ask you to change this to just `gsl::not_null...` as in, remove the const. Same on lines 35 and 49below.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Looks good! A few small comments you can squash right away when you rebase

**Comment by ffoucart** [comment]
@nilsdeppe More changes than I'd have liked in order to match the latest clang-tidy requirements, but this should be ready for you again. Note that I needed to add a default constructor to the packet struct -- clang-tidy was unhappy with packet creation that did not explicitly set the contents of the packet.

**Comment by ffoucart** [comment]
(The failing test is some paraview thing... I'm guessing this has nothing to do with this PR)

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
CI failure is some unrelated paraview thing.

**Pull request merged**